edx_proctoring/signals.py

```diff
@@ -105,6 +106,18 @@ def on_attempt_changed(sender, instance, signal, **kwargs):  # pylint: disable=u
             # to archive it
             original = sender.objects.get(id=instance.id)
 
+            # if the exam was finished for the first time, we want to mark it as
+            # complete in the Completion Service.
+            # This functionality was added because regardless of submission status
+            # on individual problems, we want to mark the entire exam as complete
+            # when the exam is finished since there are no more actions a learner can take.
+            if not original.completed_at and instance.completed_at:
```

Comment:
This approach looks good to me, but I wanted to point out that a learner could have multiple attempts for a single exam (this is due to exams being "resumable" now). 

However, if a learner does have multiple attempts only one should have a `completed_at` field - this is because we only mark the time for `completed_at` if an attempt transitions to `submitted`, and a learner cannot resume an exam once they have a `submitted` status on an attempt. 

Just wanted to point this out in case this would need to be triggered by anything other than a status change to `submitted` (i.e. if instead an attempt should be considered completed once it is reviewed)

Reply:
Thank you for checking in on this! I believe I do want just the scenario where they have submitted so I think we are all good.
The intent here is that as soon as a learner has no more actions they can do, we mark the whole exam as complete. This is to help account for cases where a learner leaves a question blank and then the entire exam is forever incomplete, but the learner has no ability to correct for it either.